Создана возможность добавление книги в избранное

k3ssL opened the pull request.

k3ssL committed:
api файл

k3ssL committed:
создан user favorite api

k3ssL committed:
созданы интерфейсы запросов / ответов, создана модель категорий

k3ssL committed:
написан user favorite service, написаны модели DTO

k3ssL committed:
fix transition

k3ssL committed:
изменены интерфейсы

k3ssL committed:
добавлена иконка закладки

k3ssL committed:
ренейм

BlackSoulHub commented on src/entities/UserFavorite/service/UserFavoriteService.ts:
```diff
@@ -0,0 +1,32 @@
+import {IShortUserFavoriteDTO} from "@/entities/UserFavorite/models/dto/IShortUserFavoriteDTO";
+
+export class UserFavoriteService {
+    static async addBookInFavorite(requestData: IAddBookInFavoriteRequest): Promise<EmptyResult> {
```
request в сервисах. Можешь просто разложить его на отдельные параметры

BlackSoulHub commented on src/pages/MainPage/UI/BookPreview/BookPreview.tsx:
```diff
@@ -9,12 +12,17 @@ interface IBookPreview {
 
-const BookPreview: React.FC<IBookPreview> = ({ image, name, author }) => {
+const BookPreview: React.FC<IBookPreview> = ({ id, image, name, author }) => {
+    const { handleHover, hover } = useBookPreview(id, "InPlans")
```
Судя по всему ты его не залил

BlackSoulHub replied:
Несколько вопросов:
1. Зачем ты изменяешь `BookPreview`? Ты хочешь сделать для пользователя статус добавленной книги на главной странице? Хотя это звучит неплохо, но это - не то, что пока нужно. Тебе нужно изменить страницу книги. Роут вроде: `books/{bookId}`.
2. Помимо этого могу предположить, что `InPlans` в данном случае - стандартное значение для книги. Но отображать нужно не так. Тебе нужно отображать то, в каком статусе находится книга у конкретного пользователя. 
Стоит учитывать: 
- Пользователь может быть не авторизован (гость). Тогда статуса книги всегда нет. Пуская будет что-то вроде кнопки: `Добавить в избранное` при нажатии на которую пользователя попросят авторизоваться.
- Пользователь может НЕ иметь книгу данную книгу в избранном. Тогда можешь сделать тоже, что и с гостем, но при нажатии должен быть список возможных вариантов
- Пользователь может уже добавить данную книгу в свой список. Тогда должна быть возможность перенести книгу в ДРУГОЙ список

BlackSoulHub reviewed (changes requested):
Кстати. Заметил, что ты не соблюдаешь наши "конвенции". 
Хотя возможно их никто не объяснял.
Попробуй:
```bash
npm run custom-lint
```

или
```bash
npm run custom-lint-auto-fix
```
Но учитывай, что `custom-lint-auto-fix` не доделан и к использованию не рекомендуется. Он исправляет нейминг файлов, но сбивает импорты

k3ssL committed:
исправление

k3ssL committed:
layout

k3ssL committed:
убрана модель DTO

k3ssL committed:
создан компонент для выбора статуса, написана логика, создан хук

k3ssL committed:
рефакторинг

k3ssL committed:
ренейм

k3ssL committed:
layout

BlackSoulHub commented on src/entities/UserFavorite/api/UserFavoriteApi.ts:
```diff
@@ -0,0 +1,51 @@
+
+    static async getBookInFavoriteById(bookId: number): Promise<Result<IGetBookInFavoriteByIdResponse>> {
+        const result = await new HTTPRequest<IGetBookInFavoriteByIdResponse>()
+            .withUrl(`/v1/ysers/favorite/my/book/${bookId}`)
```
url

BlackSoulHub commented on src/entities/UserFavorite/models/dto/IAddUserFavoriteDTO.ts:
```diff
@@ -0,0 +1,6 @@
+import {UserBookStatus} from "@/entities/UserFavorite/UserBookStatuses";
+
+export interface IAddUserFavoriteDTO {
```
Нет смысла в этом интерфейсе. Используй сигнатуру функции:

```ts
async function addUserFavoriteAsync(bookId: number, status: UserBookStatus): Promise<EmptyResult> {
    // logic...
}

```

BlackSoulHub commented on src/entities/UserFavorite/models/dto/IShortUserFavoriteDTO.ts:
```diff
@@ -0,0 +1,12 @@
+import { UserBookStatus } from "@/entities/UserFavorite/UserBookStatuses"
+
+export interface IShortUserFavoriteDTO {
```
Название модели в единственном числе. При это представляет собой модель - список книг, добавленных пользователем.

BlackSoulHub commented on src/entities/UserFavorite/models/dto/IUserFavoriteDTO.ts:
```diff
@@ -0,0 +1,13 @@
+import { UserBookStatus } from "@/entities/UserFavorite/UserBookStatuses"
+
+export interface IUserFavoriteDTO {
```
Модель короче чем, `IShortUserFavoriteDTO`. Всё те же вопросы

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/BookImage/BookImage.tsx:
```diff
@@ -1,12 +1,19 @@
 
-const BookImage: FC<IBookImageProps> = ({ image }) => {
+const BookImage: FC<IBookImageProps> = ({ image, bookId }) => {
+    const [selectedStatus, setSelectedStatus] = useState<UserBookStatus | null>(null)
```
логика в компоненте

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/BookImage/BookImage.tsx:
```diff
@@ -21,6 +28,11 @@ const BookImage: FC<IBookImageProps> = ({ image }) => {
             <div className={classes.buttons}>
                 <Button className={classes.button}>Продолжить читать</Button>
                 <Button className={classes.button}>Добавить в список</Button>
+                <SelectStatusButton
```
Ты сделал кнопку для добавления книги в список пользователя. Удали старую кнопку: `Добавить в список`

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/SelectStatusButton/useSelectStatusButton.ts:
```diff
@@ -0,0 +1,45 @@
+        setIsOpen((prevState) => !prevState)
+    }
+
+    function isLogIn(): boolean {
```
Вынеси временно в `TokenService`

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/SelectStatusButton/useSelectStatusButton.ts:
```diff
@@ -0,0 +1,45 @@
+
+        return true
+    }
+    async function ClickStatusHandler(status: UserBookStatus) {
```
Пробел

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/SelectStatusButton/useSelectStatusButton.ts:
```diff
@@ -0,0 +1,45 @@
+        })
+
+        if (!isLogIn()) {
+            window.location.href = "/auth"
```
Замени на `RedirectService`

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/SelectStatusButton/useSelectStatusButton.ts:
```diff
@@ -0,0 +1,45 @@
+            return AlertService.errorMessage(result.getError().errorMessage)
+        }
+
+        return AlertService.successMessage("Книга была добавлена")
```
После успешного добавления текущий статус книги не обновляется?

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/SelectStatusButton/SelectStatusButton.tsx:
```diff
@@ -0,0 +1,38 @@
+            </div>
+            <div className={classes.selected_status}>
+                {selectedStatus
+                    ? <p>Текущий статус: {selectedStatus}</p>
```
Это не должно быть в виде отдельного блока. Сделай подпись на кнопке. В случае, если книга не добавлена в список пользователя напиши: "Добавить в список"

k3ssL committed:
url

k3ssL committed:
добавлена сигнатура функции

k3ssL committed:
ренейм

k3ssL committed:
убрана логика в компоненте, убрана лишняя кнопка

k3ssL committed:
убрана лишняя кнопка

k3ssL committed:
метод вынесен в token service

k3ssL committed:
создан redirect servic

BlackSoulHub committed:
merge dev

# Conflicts:
#	src/app/books/layout.tsx
#	src/pages/MainPage/UI/BookPreview/BookPreview.module.css
#	src/pages/MainPage/UI/BookPreview/BookPreview.tsx
#	src/shared/UI/button/Button.module.css

BlackSoulHub committed:
add display pretty book statuses, fix key waring

BlackSoulHub committed:
style fix

BlackSoulHub committed:
remove author

BlackSoulHub committed:
add method for check if user admin

BlackSoulHub committed:
create method for redirect to auth page

k3ssL committed:
реализован вывод статуса книги

k3ssL commented on src/shared/UI/button/Button.tsx:
```diff
@@ -2,7 +2,7 @@ import React from "react"
 
 interface IButtonProps {
-    children: string
+    children: any
```
насчет этого момента.
Я не знаю какой тип данных здесь поставить, потому что компонент уже используется и изменения приведут к поломке

BlackSoulHub replied:
```suggestion
    children: string
```

```tsx
<Button onClick={ClickHandler} styles={{margin: "15px"}}>
    {selectedStatus
        ? getPrettyBookStatusName(selectedStatus)
        : "Добавить в список"
    }
</Button>
```

BlackSoulHub committed:
refactoring, style fix, fix display book status

BlackSoulHub commented on src/pages/BookPages/BookPage/UI/SelectStatusButton/SelectStatusButton.tsx:
```diff
@@ -0,0 +1,36 @@
+        <div className={classes.wrapper}>
+            <Button onClick={ClickHandler} styles={{margin: "15px"}}>
+                {selectedStatus
+                    ? <p>{getPrettyBookStatusName(selectedStatus)}</p>
```
Кнопка должна быть такого-же размера как и верхняя кнопка. Так выглядит странно
![image](https://github.com/ProjectMeowInc/meowlib.frontend.remaster/assets/86704101/06447aa4-13bc-436a-9a81-e9e5133fc4fa)

k3ssL committed:
изменены размеры кнопки

k3ssL committed:
Merge remote-tracking branch 'origin/feature/user-favorite' into feature/user-favorite

Lx1Cx committed:
Добавлена адаптация для book image

Lx1Cx committed:
Были немного переписаны стили для компонента

Lx1Cx committed:
Добавлена иконка стрелки
